platform/particle_electron/export/aux/uart_bus.hpp

```diff
@@ -0,0 +1,246 @@
+{
+
+//! Represents distinct peripheral devices
+enum class uart_device
+{
+    serial0,
+    serial1,
+    serial2,
+    serial4,
+    serial5,
+};
+
+//! \brief Electron UART bus wrapper
+template<uart_device dev>
+class uart_bus
+{
+public:
+    // Convenient type aliases.
+    using channel       = ecl::bus_channel;
+    using event         = ecl::bus_event;
+    using handler_fn    = ecl::bus_handler;
+
+    // Static interface only
+    uart_bus() = delete;
+    ~uart_bus() = delete;
+
+    // Should not be copied.
+    uart_bus &operator=(uart_bus&) = delete;
+    uart_bus(uart_bus&) = delete;
+
+    //! \brief Lazy initialization.
+    //! \return Status of operation.
+    static err init();
+
+    //! \brief Sets rx buffer with given size.
+    //! \param[in,out]  rx      Buffer to write data to. Optional.
+    //! \param[in]      size    Size
+    static err set_rx(uint8_t *rx, size_t size);
+
+    //! \brief Sets tx buffer made-up from sequence of similar bytes.
+    //! \param[in] size         Size of sequence
+    //! \param[in] fill_byte    Byte to fill a sequence. Optional.
+    static err set_tx(size_t size, uint8_t fill_byte = 0xff);
+
+    //! \brief Sets tx buffer with given size.
+    //! \param[in] tx   Buffer to transmit. Optional.
+    //! \param[in] size Buffer size.
+    static err set_tx(const uint8_t *tx, size_t size);
+
+    //! \brief Sets event handler.
+    //! Handler will be used by the bus, until reset_handler() will be called.
+    //! \param[in] handler Handler itself.
+    static err set_handler(const handler_fn &handler);
+
+    //! \brief Reset xfer buffers.
+    //! Buffers that were set by \sa set_tx() and \sa set_rx()
+    //! will be no longer used after this call.
+    static err reset_buffers();
+
+    //! \brief Resets previously set handler.
+    static err reset_handler();
+
+    //! \brief Executes xfer, using buffers previously set.
+    //! When it will be done, handler will be invoked.
+    //! \return Status of operation.
+    static err do_xfer();
+
+    //! \brief Cancels xfer.
+    //! After this call no xfer will occur.
+    //! \return Status of operation.
+    static err cancel_xfer();
+
+private:
+    //! Gets correct serial object for the device.
+
+    template<uart_device d = dev, typename std::enable_if_t<d == uart_device::serial0, int> = 0>
+    static constexpr USBSerial &get_serial();
+
+    //! Gets correct serial object for the device.
+    template<uart_device d = dev, typename std::enable_if_t<d != uart_device::serial0, int> = 0>
+    static constexpr USARTSerial &get_serial();
+
+    static const uint8_t                *m_tx;           //! Transmit buffer.
+    static size_t                       m_tx_size;       //! TX buffer size.
+    static uint8_t                      *m_rx;           //! Receive buffer.
+    static size_t                       m_rx_size;       //! RX buffer size.
+
+    static safe_storage<handler_fn>     m_ev_storage;    //! Storage for event handler.
```

Comment:
I propose to rename this to `m_ev_handler`. I find this more informative since we're storing the event handler, not the event.